MINIBALL: CPU virtualization

yaoyinnan opened the pull request:
Refer to the vmm-reference project to implement Miniball, and use dragonball-sandbox creats to implement CPU virtualization.

Add description document for CPU virtualization part. Modify the description of CPU virtualization in MiniBall's README.md.

Fixes: https://github.com/openanolis/dragonball-sandbox/issues/181

ASOC: https://github.com/openanolis/dragonball-sandbox/issues/155

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

studychao commented:
Please remember to push the document for CPU virtualization (Chinese one is ok)

Maybe in other PR.

studychao commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,736 @@
+        {
+            vcpu.configure_cpuid(&vcpu.config.cpuid)?;
+            vcpu.configure_msrs()?;
+            // vcpu.configure_regs(memory)?;
```
Why this is commented out?

yaoyinnan replied:
It has been removed, and the reg should be set at run().

studychao replied:
Got that

studychao commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,736 @@
+        Ok(vcpu)
+    }
+
+    /// Set CPUID.
```
Usually, only `pub fn` need `///` ([Doc comments](https://doc.rust-lang.org/reference/comments.html#doc-comments))

yaoyinnan replied:
Got it, updated it

yaoyinnan commented:
@studychao I'm still revising the CPU virtualization docs, which will be submitted in other PR in the future.

studychao commented:
> @studychao I'm still revising the CPU virtualization docs, which will be submitted in other PR in the future.

Got that, looking forward to that.

wllenyj commented on crates/dbs-arch/Cargo.toml:
```diff
@@ -15,7 +15,7 @@ memoffset = "0.6"
 kvm-bindings = { version = "0.5.0", features = ["fam-wrappers"] }
 kvm-ioctls = ">=0.9.0"
 vm-memory = { version = "0.7" }
-vmm-sys-util = "0.9.0"
```
Can this change be deleted?

yaoyinnan replied:
If I don't modify it, there will be compile errors. I'll remove this change later when the compilation errors are resolved. Ultimately I'll promise not to affect other dbs crates.

studychao commented on crates/dbs-miniball/src/vmm/src/boot.rs:
```diff
@@ -0,0 +1,267 @@
+    MmioGapStartPastMmioGapEnd,
+}
+
+fn add_e820_entry(
```
This function is provided in dbs-boot (or you didn't start to change this part?)

yaoyinnan replied:
I plan to modify it in memory virtualization.

studychao replied:
Got that

studychao commented:
Could you declear what file you have changed?

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vm.rs:
```diff
@@ -0,0 +1,662 @@
+        let mut vm = Self::create_vm(kvm, vm_config, exit_handler, guest_memory)?;
+
+        #[cfg(target_arch = "x86_64")]
+        mptable::setup_mptable(guest_memory, vm.config.num_vcpus, vm.config.num_vcpus)
```
Use setup_mptable() from dbs-boot to set MPTable.

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+            let mut cpuid = base_cpuid.clone();
+
+            #[cfg(target_arch = "x86_64")]
+            let vm_spec = VmSpec::new(0, 2, 1, 1, 1, VpmuFeatureLevel::Disabled).unwrap();
```
Use VmSpec and process_cpuid() to filter and set CPUID.

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+            .map_err(Error::KvmIoctl)?;
+
+        #[cfg(target_arch = "x86_64")]
+        let supported_msr_list =
```
Because the `supported_guest_msrs()` only return `MsrList`. I can only implement the function to get `Msrs` here.

`supported_guest_msrs()` in dbs-arch:
```
pub fn supported_guest_msrs(kvm_fd: &Kvm) -> Result<MsrList> {
    let mut msr_list = kvm_fd
        .get_msr_index_list()
        .map_err(Error::GetSupportedModelSpecificRegisters)?;

    msr_list.retain(|msr_index| msr_should_serialize(*msr_index));

    Ok(msr_list)
}
```

`supported_guest_msrs()` in vmm-reference:
```
pub fn supported_guest_msrs(kvm_fd: &Kvm) -> Result<Msrs> {
    let mut msr_list = kvm_fd
        .get_msr_index_list()
        .map_err(Error::GetSupportedMSR)?;

    msr_list.retain(|msr_index| msr_should_serialize(*msr_index));

    let mut msrs =
        Msrs::new(msr_list.as_fam_struct_ref().nmsrs as usize).map_err(|_| Error::CreateMsrs)?;
    let indices = msr_list.as_slice();
    let msr_entries = msrs.as_mut_slice();
    // We created the msrs from the msr_list. If the size is not the same,
    // there is a fatal programming error.
    assert_eq!(indices.len(), msr_entries.len());
    for (pos, index) in indices.iter().enumerate() {
        msr_entries[pos].index = *index;
    }

    Ok(msrs)
}
```

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+    // Configure MSRs.
+    #[cfg(target_arch = "x86_64")]
+    fn configure_msrs(&self) -> Result<()> {
+        dbs_arch::regs::setup_msrs(&self.vcpu_fd).map_err(Error::MSRSConfiguration)
```
Use setup_msrs() from dbs-adch to set msrs.

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+    // Configure regs.
+    #[cfg(target_arch = "x86_64")]
+    fn configure_regs(&self, instruction_pointer: GuestAddress) -> Result<()> {
+        dbs_arch::regs::setup_regs(
```
Use setup_regs() from dbs_arch to set regs.

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+
+    // Configure sregs.
+    #[cfg(target_arch = "x86_64")]
+    fn configure_sregs<M: GuestMemory>(&self, guest_memory: &M) -> Result<()> {
```
Use setup_sregs() from dbs-arch to set sregs.

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+    // Configure FPU.
+    #[cfg(target_arch = "x86_64")]
+    fn configure_fpu(&self) -> Result<()> {
+        dbs_arch::regs::setup_fpu(&self.vcpu_fd).map_err(Error::FPUConfiguration)
```
Use setup_fpu() from dbs-arch to set fpu.

yaoyinnan commented on crates/dbs-miniball/src/vm-vcpu/src/vcpu/mod.rs:
```diff
@@ -0,0 +1,735 @@
+    // Configures LAPICs. LAPIC0 is set for external interrupts, LAPIC1 is set for NMI.
+    #[cfg(target_arch = "x86_64")]
+    fn configure_lapic(&self) -> Result<()> {
+        dbs_arch::interrupts::set_lint(&self.vcpu_fd).map_err(Error::LocalIntConfiguration)
```
Use set_lint() from dbs-arch to set lapic.

yaoyinnan reviewed (commented):
@studychao I comment what file I have changed.

yaoyinnan commented:
@studychao I submitted a CPU virtualization description document, please help me review.

studychao commented:
> @studychao I submitted a CPU virtualization description document, please help me review.

Could you create a separate Pr for that？

yaoyinnan commented:
> > @studychao I submitted a CPU virtualization description document, please help me review.
> 
> Could you create a separate Pr for that？

OK, I created a separate Pr: https://github.com/openanolis/dragonball-sandbox/pull/188.

studychao commented:
Since you have solved compile problem for dbs-arch, is it ok to add miniball to the Dragonball-sandbox workspace?

yaoyinnan commented:
> Since you have solved compile problem for dbs-arch, is it ok to add miniball to the Dragonball-sandbox workspace?

@studychao 
Due to the need to specify the version of vmm-sys-utils, I modified Cargo.lock in the root directory of dragonball-sandbox and included it in the commit.

Also, since dbs-miniball needs to depend on a specific version of linux-loader (9a9f071), and after dbs-miniball is added to the workspace, the patch in the crate of project will no longer take effect. I added patch to Cargo.lock in the root directory:

````
[patch.crates-io]
# TODO: Using this patch until a version > 4.0 gets published.
linux-loader = { git = "https://github.com/rust-vmm/linux-loader.git", rev = "9a9f071" }
````

yaoyinnan committed:
MINIBALL: CPU virtualization

Refer to the vmm-reference project to implement Miniball, and use dragonball-sandbox creats to implement CPU virtualization.

Fixes: https://github.com/openanolis/dragonball-sandbox/issues/181

Signed-off-by: yaoyinnan <yaoyinnan@foxmail.com>

studychao reviewed (approved):
LGTM:)
